From ddb739a87b015e98a8d7cbc6250da832235262d8 Mon Sep 17 00:00:00 2001
From: Pawel Moll <pawel.moll@st.com>
Date: Wed, 24 Sep 2008 17:25:58 +0100
Subject: [PATCH] sound_stm: Frequency adjustment control is keeping the setting now.

Oversampling frequency adjustment value was cleared on player
configuration, so the settings were lost (amongst others) in case
of xruns. Now they are persistent, as intuition says...

Signed-off-by: Pawel Moll <pawel.moll@st.com>
---
 sound/stm/fsynth.c |   12 ++++++------
 1 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/sound/stm/fsynth.c b/sound/stm/fsynth.c
index 8ca51f1..247390e 100644
--- a/sound/stm/fsynth.c
+++ b/sound/stm/fsynth.c
@@ -424,7 +424,6 @@ static int snd_stm_fsynth_adjustment_get(struct snd_kcontrol *kcontrol,
 static int snd_stm_fsynth_adjustment_put(struct snd_kcontrol *kcontrol,
 		struct snd_ctl_elem_value *ucontrol)
 {
-	int result;
 	struct snd_stm_fsynth_channel *fsynth_channel =
 		snd_kcontrol_chip(kcontrol);
 	struct snd_stm_fsynth *fsynth;
@@ -445,11 +444,12 @@ static int snd_stm_fsynth_adjustment_put(struct snd_kcontrol *kcontrol,
 	channel = fsynth_channel - fsynth_channel->fsynth->channels;
 	old_adjustement = fsynth_channel->adjustment;
 
-	result = snd_stm_fsynth_channel_configure(fsynth, channel,
+	/* If the synthesizer hasn't been configured yet... */
+	if (fsynth_channel->frequency == 0)
+		fsynth_channel->adjustment = ucontrol->value.integer.value[0];
+	else if (snd_stm_fsynth_channel_configure(fsynth, channel,
 			fsynth_channel->frequency,
-			ucontrol->value.integer.value[0]);
-
-	if (result < 0)
+			ucontrol->value.integer.value[0]) < 0)
 		return -EINVAL;
 
 	return old_adjustement != fsynth_channel->adjustment;
@@ -512,7 +512,7 @@ int snd_stm_fsynth_set_frequency(struct snd_stm_fsynth_channel *fsynth_channel,
 
 	return snd_stm_fsynth_channel_configure(fsynth_channel->fsynth,
 			fsynth_channel - fsynth_channel->fsynth->channels,
-			frequency, 0);
+			frequency, fsynth_channel->adjustment);
 }
 
 int snd_stm_fsynth_add_adjustement_ctl(
-- 
1.5.3.6

